005-tailwindcss-framework.md

```diff
@@ -0,0 +1,126 @@
+
+We do appreciate having a library approach to our front-end to help with establishing a system and a shared baseline of patterns, but in practice a CSS library that has simpler updates or just requires less updates while still supporting changing design needs would be an improvement.
+
+Additionally, as our team structure and focus has changed from dedicated back-end and front-end developers to full-stack developers, we have less team members with extensive, in-depth CSS knowledge, experience or interest in writing cross-browser, responsive style rules. However, our full-stack team still needs to tackle development of features that include full styling of interfaces and showcases the need for a styling framework that is flexible, customizable and empowers our developers to confidently apply styles without the need to code custom style declarations themselves.
+
+## Proposal
+
+A Utility CSS Framework is a system that provides low-level utility classes that typically map to a corresponding single style rule. They enable you to build custom interfaces entirely with these quick and easy to write classes in the HTML for an interface, without the need to context-switch languages back-and-forth between HTML and CSS.
+
+Additionally, by using utility classes to style HTML, it has the added benefit of abstracting away specificity issues and eliminating problems with the global namespace that can cause style conflicts.
+
+By utilizing a Utility CSS Framework, and more specifically the TailwindCSS Framework, the tech team can work more confidently when building the front-end for different interfaces across our platforms.
+
+A utility class framework like Tailwind CSS offers the following benefits:
+
+- Establishes an API of class names for DoSomething's design style guide.
+- Easily style components and pages by applying a pre-configured library of classes derived from our established style guide directly to HTML.
+- Simplify styling components by significantly reducing the need to write custom CSS, and no more worrying about coming up with unique [BEM-style class names](http://getbem.com/introduction/) as designs change or new designs are developed.
+- Utility classes use functional naming (describe what they do) rather than semantic naming for composed classes, making it easier for full-stack developers to understand what styles are being applied directly to the HTML rather than having to reference a stylesheet and see what series of styles a semantic class is applying.
+- Limits the growth of CSS styles once our collection of utility classes is established and using a library like [PurgeCSS](https://www.purgecss.com/) reduces CSS size further by removing unused utility class definitions in an application.
```

Comment:
> Limits the growth of CSS styles once our collection of utility classes is established

Could you expand on this piece? Does using a Utility CSS Framework just amount to writing less CSS in general?

Reply:
Yep! This might be easier to explain with an example:

```css
.callout  {
    background-color: white;
    border: 1px solid gray;
    border-radius: 4px;
    display: block;
    font-size: 22px;
}

.card  {
    background-color: white;
    border: 1px solid gray;
    border-radius: 4px;
    display: block;
    font-size: 18px;
}
```

Some of the above styles could get repeated multiple times if we add another block-like composed class like `.media`, which may be slightly different but have a few things that overlap. We could try and break each down farther for some sort of very generic class that has just the `background-color`, `display` and `border` definitions, but then you'd have the CSS that builds those components out defined in different files. Makes it much harder to maintain and harder to keep track of if we don't have a dedicated front-end team. Some of these effects we see now in our current approaches.

With a utility css framework, the classes you'd have for the above would be the following and they would only really need to get defined _once_:

```css
.bg-white {
    background-color: white;
}

.border {
    border-width: 1px;
}

.border-gray-200 {
    border-color: #edf2f7;
}

.border-solid {
    border-style: solid;
}

.rounded {
    border-radius: .25rem;
}

.block {
    display: block;
}

.text-base {
    font-size: 1rem;
}

.text-lg {
    font-size: 1.125rem;
} 
```

In contrast, I did a quick search on Phoenix and `display: block;` is used 31 times (kind of surprised it's only that, likely because we started using some utility classes). Once tailwind is implemented, it would ideally only be defined once and then the class that defines it `.block`, would be using within the HTML only.

So once we have our set of utility classes established, our CSS shouldn't really grow much over time since we'll be building by just reusing class names to define styles.

Reply:
THANK YOU this example is extremely helpful!! Thank you for taking the time to spell out a real world example, really helped me understand it a lot better!

Reply:
+1, this is a great example!